runtime/src/main/java/org/corfudb/runtime/object/transactions/AbstractTransactionalContext.java

```diff
@@ -124,6 +125,13 @@ public abstract class AbstractTransactionalContext implements
     @Setter
     private TxnContext txnContext;
 
+    /**
+     * Flag used to track if this transaction has performed any accesses
+     * on monotonic objects. This is used to compute the commit address
+     * of read-only transactions.
+     */
+    protected boolean hasAccessedMonotonicObject = false;
```

Comment:
Now you could just add a statement in `setUncommittedChanges`:

```
            if (isMonotonicObject()) {
                TransactionalContext.getCurrentContext().hasAccessedMonotonicObject = true;
            }
```

I'll let you decide if you want to go down this route. To me, passing around `ICorfuSMRProxyInternal` seems a bit awkward, as the object has a large scope.